### ISSUE TITLE
builder 0.14.10 regression 

### ISSUE THREAD

**Issue by typhonrt** [systemjs/builder#384]
Greets Guy et al,

I'm still trying to track things down, but a complicated (to me) side effect / regression affects the changes implemented for #379 between `0.14.9` and `0.14.10`. I have so far confirmed I can apply the #379 patch to 0.14.9 and everything works as expected. Running against 0.14.10 the build completes fine, but there is a runtime error. I have verified that the code surrounding #379 seems to be working correctly in 0.14.10, so my only current assumption is that there is another side effect / regression. #379 was in builder.js and I have validated that the regression still exists when swapping builder.js (0.14.9) with #379 applied to 0.14.10. It seems the problem lies with other changes beyond builder.js as things go, so is complicated for me to debug though I am going to continue investigating. 

UPDATE / BELOW is related to #386:
`------`
The error that I'm seeing in the static build is that it completes and on load it doesn't seem to be loading correctly the user mapped paths which depend on a previous config load. IE see the code example in #377 that should work. I am now examining a folder compare between 0.14.9 and 0.14.10. 

It should be noted that everything works fine when in the browser via full source, so definitely a builder issue. In my framework code for a web app an import of custom mapped paths that should be loaded and executed simply isn't. In this case it is initialization code for Parse, so the error happens quickly on load. Due to hoisted ES6 imports and / or how SystemJS Builder works it's hard to debug.
`------`

There is however an actual regression here described in follow up comments. It turns out that using `loadConfig` for additional config loading when queueing multiple Builder / bundle operations causes a failure / clobbered configuration data. More is described below.

Guy, could you offer any insight into any conceivable changes between these versions? I'm pretty sure the regression may be beyond `builder.js`. I'm fishing right now... 

Thanks...

**Comment by typhonrt** [comment]
OK.. I think I've narrowed down the trigger condition. It turns out that the problem I'm seeing is due to my Gulp task which queues multiple bundle / build operations. Something is being reset / order of operations different internally to SystemJS Builder in 0.14.10 versus 0.14.9.  

BTW I have compiled a NPM package, [typhon-core-gulptasks](https://www.npmjs.com/package/typhon-core-gulptasks) of common Gulp tasks for use with JSPM / SystemJS. It definitely helps to share workflows across many projects and is really easy to import into projects. 

Here is the current Gulp task and the build function being executed per bundle op. Bundle ops are defined in a separate JSON file like this one ([bundle-config.json](https://github.com/typhonjs/backbone-parse-es6/blob/master/config/bundle-config.json)):
https://github.com/typhonjs/typhon-core-gulptasks/blob/master/tasks/jspm.js#L50
https://github.com/typhonjs/typhon-core-gulptasks/blob/master/tasks/jspm.js#L160

In particular this is likely due to simultaneous build operations being kicked off via promises versus a single build operation or multiple running synchronously. As mentioned if I only queue up one bundle / build operation everything works as expected. Queue up more than one at a time and things break which is the "hidden regression" that I'm trying to work around or learn more about. 

In the `buildStatic` function above I do create a new Builder per bundle / build operation wrapped in a Promise followed up by Promise.all() in the Gulp task.

```
return new Promise(...
{
   var builder = new jspm.Builder('./config.js');
   ...
   builder.buildStatic(...);
}
```

Since my Gulp task is the one that kicks off this pressure point I'm looking at reworking the task first before deeply investigating internals to SystemJS Builder. It might be nice to know what might influence the change if you have any insights on changes to Builder which might make it fragile to the code I've listed above. 

Thanks for any help!

**Comment by typhonrt** [comment]
Even more details... 

It seems like there may be a regression issue with `loadConfig` when multiple Builder operations are kicked off. 

if I change this:
```
      var builder = new jspm.Builder('./config.js');
      builder.loadConfig(additionalConfig).then(function()
      {
         // If extraConfig is an object literal / hash then load it now.
         if (typeof extraConfig === 'object')
         {
            builder.config(extraConfig);
         }
         ...
```

To using the synchronized loadConfigSync for the follow up additional config file then things work fine when queueing multiple bundle / build ops:

```
      var builder = new jspm.Builder('./config.js');

      var extraConfigType = typeof extraConfig;
      if (extraConfigType === 'string')
      {
         builder.loadConfigSync(extraConfig)
      }
      else if (extraConfigType === 'object')
      {
         builder.config(extraConfig);
      }
```

Since things worked with SystemJS Builder `0.14.9` and not `0.14.10` it seems to be an actual regression. However, I have a solution for my needs / Gulp task now by using `loadConfigSync` for any additional config loading.

**Comment by fictitious** [comment]
Yes I'm also seeing this: 

        var Builder = require('systemjs-builder');
        var builder = new Builder;
        builder.loadConfig('config.js');
        builder.bundle( ... );

        var builder2 = new Builder;
        builder2.loadConfig('config.js');
        builder2.bundle( ... );

gives an error which I traced to the first builder having empty config at the time its bundle() method is called. Interestingly, it happens on slower computer only.

I would not say it's quite unexpected however, given that  loadConfig() and bundle() are asynchronous, and all builders are using global loader instance (which kind of makes sense because probably there would be only one instance of the loader when it gets implemented in the browser).

This works:

        var Builder = require('systemjs-builder');
        var builder = new Builder;
        builder.loadConfig('config.js').then(function() {
            builder.bundle( ... ).then(function() {

                var builder2 = new Builder;
                builder2.loadConfig('config.js').then(function() {
                    builder2.bundle( ... );
                });
            });
        });

I'd say it's not a bug.

**Comment by typhonrt** [comment]
Hard to say on it being a bug or possible regression at least as I can load up Builder 0.14.9 and version before and parallel builder usage w/ loadConfig works; 0.14.10 same code failed. 

I did find a workaround for my use case though which is loadConfigSync as in my case I'm already wrapping things in a promise around the entire builder instantiation / buildStatic call to queue up many separate bundle tasks.

**Comment by guybedford** [comment]
It would be useful to know what config operations are applying here. In the example case of:

```javascript
   var builder = new jspm.Builder('./config.js');

   builder.loadConfig('./config/config-app-paths.js').then(function()
   {
      ...
   }
```

Is it the paths in `config.js` or `config-app-paths.js` which aren't applying? Are you always instantiating a new builder instance for every task, or do you ever run `builder.reset()`?

**Comment by typhonrt** [comment]
Hi Guy... I have solved the problem by using `loadConfigSync` for loading additional config data.

In the Gulp task 6 bundle operations are being queued up in rapid succession using separate Builder instances and no usage of `builder.reset()`. `./config.js` is the standard JSPM config of installed packages from `package.json`. `config-app-paths.js` had additional `map` entries, but also references `System.map`. 

It could be that `paths` was not applying or somehow out of sync. What I saw was that a package from `./config.js` that should resolve in successive queueing of bundle operations was not being resolved.

This did work prior to `0.14.10` and I've been using the Gulp task for over a month, so previous Builder versions worked as well.

**Comment by fictitious** [comment]
I have a failing testcase with the code along these lines:

        var builder1 = new Builder();
        var builder2 = new Builder();

        builder1.loadConfig('config1.js').then(function() { ... });
        builder2.loadConfig('config2.js').then(function() { ... });

Is this supposed to work? If yes, I have two-part PR  #394   with the first commit adding the testcase (that reliably fails for me), and the second commit with one possible way to fix it. I tried to explain why it fails in the comment for the second commit.

**Comment by guybedford** [comment]
This should be fixed by the merge of https://github.com/systemjs/builder/commit/733eccf478f207a55ac7c2ed0b3b6ecffb6c7244.

**Comment by guybedford** [comment]
Released in 0.14.11.
